api/src/main/java/org/openmrs/util/ConceptReferenceRangeUtility.java

```diff
@@ -0,0 +1,143 @@
+			throw new IllegalArgumentException("Failed to evaluate criteria with reason: patient is null");
+		}
+		
+		if (StringUtils.isBlank(criteria)) {
+			throw new IllegalArgumentException("Failed to evaluate criteria with reason: criteria is empty");
+		}
+		
+		VelocityContext velocityContext = new VelocityContext();
+		velocityContext.put("fn", this);
+		velocityContext.put("obs", obs);
+		
+		velocityContext.put("patient", obs.getPerson());
+		
+		VelocityEngine velocityEngine = new VelocityEngine();
+		
+		StringWriter writer = new StringWriter();
+		String wrappedCriteria = "#set( $criteria = " + criteria + " )$criteria";
+		
+		try {
+			velocityEngine.evaluate(velocityContext, writer, ConceptReferenceRangeUtility.class.getName(), wrappedCriteria);
+			return Boolean.parseBoolean(writer.toString());
+		}
+		catch (ParseErrorException e) {
+			throw new APIException("An error occurred while evaluating criteria: Invalid criteria: " + criteria, e);
+		}
+		catch (Exception e) {
+			throw new APIException("An error occurred while evaluating criteria: ", e);
+		}
+	}
+	
+	/**
+	 * Gets the latest Obs by concept.
+	 *
+	 * @param conceptRef can be either concept uuid or conceptMap's code and sourceName 
+	 *                   e.g "bac25fd5-c143-4e43-bffe-4eb1e7efb6ce" or "CIEL:1434"
+	 * @param person person to get obs for
+	 *                   
+	 * @return Obs
+	 */
+	public Obs getLatestObs(String conceptRef, Person person) {
+		Concept concept = Context.getConceptService().getConceptByReference(conceptRef);
+
+		if (concept != null) {
+			List<Obs> observations = Context.getObsService().getObservations(
+				Collections.singletonList(person), 
+				null, 
+				Collections.singletonList(concept), 
+				null, 
+				null, 
+				null,
+				Collections.singletonList("dateCreated"), 
+				1, 
+				null,
+				null, 
+				null, 
+				false
+			);
+
+			return observations.isEmpty() ? null : observations.get(0);
+		}
+
+		return null;
+	}
+	
+	/**
+	 * Gets the time of the day in hours.
+	 *
+	 * @return the hour of the day in 24hr format (e.g. 14 to mean 2pm)
+	 */
+	public int getCurrentHour() {
+		return LocalTime.now().getHourOfDay();
+	}
+	
+	/**
+	 * Retrieves the most relevant Obs for the given current Obs and conceptRef. If the current Obs contains a valid value 
+	 * (coded, numeric, date, or text) and the concept in Obs is the same as the supplied concept,
+	 * the method returns the current Obs. Otherwise, it fetches the latest Obs for the supplied concept and patient.
+	 *
+	 * @param currentObs the current Obs being evaluated
+	 * @return the most relevant Obs based on the current Obs, or the latest Obs if the current one has no valid value
+	 */
+	public Obs getCurrentObs(String conceptRef, Obs currentObs) {
+		Concept concept = Context.getConceptService().getConceptByReference(conceptRef);
+		
+		if ((currentObs.getValueCoded() != null || currentObs.getValueNumeric() != null || currentObs.getValueDate() != null || currentObs
```

Comment:
How about the rest of the values like boolean, etc? You can see them in Obs.getValueAsString